### PR TITLE
keyless prover vk rotation zero downtime

### PR DESCRIPTION
**Pull request by zjma**
## Description

- New on-chain config: `SetupsVKs` to store (Setup, VK) pairs.
- New keyless transaction field that takes a `setup_id`.
- In keyless txn validation, if a `setup_id` is specified, use the corresponding VK from `SetupsVKs` config.

## Type of Change
- [x] Refactoring

## Which Components or Systems Does This Change Impact?
- [x] Move/Aptos Virtual Machine
- [x] Aptos Framework

## How Has This Been Tested?
new smoke test.

## Key Areas to Review
- Does this refactoring really NOT need a feature flag?

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by alinush — COMMENTED** [review]
Let's discuss alternatives first 🙏

File: aptos-move/aptos-vm/src/aptos_vm.rs [hunk 1]
```diff
@@ -211,8 +210,7 @@ pub struct AptosVM {
     pub(crate) gas_feature_version: u64,
     gas_params: Result<AptosGasParameters, String>,
     pub(crate) storage_gas_params: Result<StorageGasParameters, String>,
-    /// For a new chain, or even mainnet, the VK might not necessarily be set.
-    pvk: Option<PreparedVerifyingKey<Bn254>>,
+    keyless_config: KeylessConfig,
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/aptos_vm.rs:213]: What is the lifetime of this config? @zjma @alinush? Per epoch, per block?

File: aptos-move/aptos-vm/src/keyless_validation.rs [hunk 2]
```diff
@@ -81,6 +84,25 @@ pub(crate) fn get_groth16_vk_onchain(
     get_resource_on_chain::<Groth16VerificationKey>(resolver)
 }
 
+pub(crate) fn get_vk_map(
+    resolver: &impl AptosMoveResolver,
+) -> Result<HashMap<String, PreparedVerifyingKey<Bn254>>, VMStatus> {
+    let SetupsVKs { setups } = get_resource_on_chain::<SetupsVKs>(resolver)?;
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/keyless_validation.rs:90]: Can we use OnChainConfig trait here and use fetch_config? You should not need a resolver but can use anything that implements ConfigStorage, right?

File: aptos-move/framework/aptos-framework/sources/keyless_account.move [hunk 3]
```diff
@@ -29,7 +29,9 @@ module aptos_framework::keyless_account {
     struct Group {}
 
     #[resource_group_member(group = aptos_framework::keyless_account::Group)]
-    /// The 288-byte Groth16 verification key (VK) for the ZK relation that implements keyless accounts
+    /// The 288-byte Groth16 verification key (VK) for the ZK relation that implements keyless accounts.
+    ///
+    /// Deprecated by `SetupsVKs`.
```
Comment (georgemitenkov) [inline aptos-move/framework/aptos-framework/sources/keyless_account.move:34]: You can use #[deprecated] I think?

File: aptos-move/aptos-vm/src/keyless_validation.rs [hunk 4]
```diff
@@ -304,3 +282,108 @@ pub(crate) fn validate_authenticators(
...
+
+            match &zk_sig.proof {
+                ZKP::Groth16(groth16proof) => {
+                    // let start = std::time::Instant::now();
+                    let public_inputs_hash = get_public_inputs_hash(sig, pk, rsa_jwk, config)
+                        .map_err(|_| {
+                            // println!("[aptos-vm][groth16] PIH computation failed");
+                            invalid_signature!("Could not compute public inputs hash")
+                        })?;
+                    // println!("Public inputs hash time: {:?}", start.elapsed());
```
Comment (georgemitenkov) [inline aptos-move/aptos-vm/src/keyless_validation.rs:316]: Do we need to have these printlns and other commented out code?

**Review by alinush — COMMENTED** [review]
Converted to draft to avoid this accidentally auto-merging.

File: types/src/keyless/mod.rs [hunk 5]
```diff
@@ -80,6 +80,10 @@ macro_rules! serialize {
 pub enum EphemeralCertificate {
     ZeroKnowledgeSig(ZeroKnowledgeSig),
     OpenIdSig(OpenIdSig),
+    ZeroKnowledgeSigV2 {
+        setup_id: String,
+        zk_sig: ZeroKnowledgeSig,
+    },
```
Comment (zjma) [inline types/src/keyless/mod.rs:86]: @larry-aptos @bowenyang007 here an authenticator change is being proposed. does it break indexer?

File: aptos-move/aptos-vm/src/aptos_vm.rs (hunk 1 above)
Comment (zjma) [reply aptos-move/aptos-vm/src/aptos_vm.rs:213]: It is up to us but here we implemented it as per-block.

File: aptos-move/aptos-vm/src/keyless_validation.rs (hunk 4 above)
Comment (zjma) [reply aptos-move/aptos-vm/src/keyless_validation.rs:316]: no, will delete.

File: aptos-move/framework/aptos-framework/sources/keyless_account.move (hunk 3 above)
Comment (zjma) [reply aptos-move/framework/aptos-framework/sources/keyless_account.move:34]: what does this annotation do exactly?

**Pull request closed without merging**